src/nvim/buffer.c

```diff
@@ -1770,6 +1770,8 @@ buf_T * buflist_new(char_u *ffname, char_u *sfname, linenr_T lnum, int flags)
   buf->b_p_bl = (flags & BLN_LISTED) ? true : false;    // init 'buflisted'
   kv_destroy(buf->update_channels);
   kv_init(buf->update_channels);
+  kv_destroy(buf->update_callbacks);
+  kv_init(buf->update_callbacks);
```

Comment:
Rather than adding an extra `foo_callbacks` field for each "attach" mechanism, would it make sense to have a new internal channel type that forwards channel events to callbacks? Then this could be used to register callbacks for any event type.

Reply:
I'm not sure it does, the implementation and semantics would both be vastly different. For `ui.c`, treating internal callback as `RemoteUI` doesn't make any sense (none of the code there is needed for internal callback), having separate `LuiUI` type to implement the same interface would be much simpler. 

In this module, we could code golf away the separate array by storing callbacks as negative numbers or something. But it wouldn't make callbacks to be channels, just storing a union of both types in the same array.